packages/datadog-plugin-sharedb/src/index.js

```diff
@@ -0,0 +1,165 @@
+}
+
+/**
+ * @description Handle a "receive" event. This will be a message to the server, from the client or the same server
+ * process.
+ * @param {Tracer} tracer DataDog Tracer.
+ * @param {sharedb} config sharedb plugin configuration.
+ * @param {string} action ShareDB Message action.
+ * @param {Object} agent ShareDB Agent
+ * @param {Object} triggerContext ShareDB trigger context (internal middleware concept).
+ * @param {Function} callback Callback to continue middleware execution.
+ * @param {Function} triggerFn Function to start middleware execution.
+ * @returns {*}
+ */
+function handleReceive (tracer, config, action, agent, triggerContext, callback, triggerFn) {
+  if (triggerContext.data && triggerContext.data.a) {
+    const scope = tracer.scope()
+    const childOf = scope.active()
+    // Call the trigger function to continue the middleware chain.
+    return triggerFn.call(this, action, agent, triggerContext, function wrappedCallback (err) {
+      // When the middleware calls back into us, start a trace.
+      const actionName = getReadableActionName(triggerContext.data.a)
+      tracer.trace(
+        'sharedb.request',
+        {
+          childOf,
+          tags: {
+            'service.name': config.service || `${tracer._service}-sharedb`,
+            'span.type': 'sharedb.request',
+            'span.kind': 'client',
+            'resource.method': actionName,
+            'resource.name': getReadableResourceName(actionName, triggerContext.data.c, triggerContext.data.q)
+          }
+        },
+        (span, spanDoneCb) => {
+          if (config.hooks && config.hooks.receive) {
+            config.hooks.receive(span, triggerContext)
+          }
+          if (span) {
+            MessagesAwaitingResponse.set(triggerContext.data, {
+              span,
+              spanDoneCb
+            })
+          }
+          callback(err)
+        })
+    })
+  } else {
+    return triggerFn.call(this, action, agent, triggerContext, callback)
+  }
+}
+
+/**
+ * @description Handle a "reply" event. This will be a message to a client for a corresponding "receive" event.
+ * @param {sharedb} config sharedb plugin configuration.
+ * @param {string} action ShareDB Message action.
+ * @param {Object} agent ShareDB Agent
+ * @param {Object} triggerContext ShareDB trigger context (internal middleware concept).
+ * @param {Function} callback Callback to continue middleware execution.
+ * @param {Function} triggerFn Function to start middleware execution.
+ * @returns {*}
+ */
+function handleReply (config, action, agent, triggerContext, callback, triggerFn) {
+  const replySpanInfo = MessagesAwaitingResponse.get(triggerContext.request)
+  if (replySpanInfo) {
+    if (config.hooks && config.hooks.reply) {
+      config.hooks.reply(replySpanInfo.span, triggerContext)
+    }
+    replySpanInfo.spanDoneCb()
+    MessagesAwaitingResponse.delete(triggerContext.request)
```

Comment:
Is `triggerContext.request` here the same as the `triggerContext.data` from the `receive` action? If yes, then is the WeakMap even needed since the span and callback could simply be added to the trigger data? And if no, then doesn't this mean that this code would do nothing?

Reply:
1. It's the same object. I know that's confusion, but that can't change.
2. We can't tag these items onto the objects passed to ShareDB, as they could get serialized and sent to the client.